torch/csrc/distributed/c10d/socket.cpp

```diff
@@ -0,0 +1,818 @@
+  return setSocketFlag(SOL_SOCKET, SO_REUSEADDR, true);
+}
+#endif
+
+std::uint16_t SocketImpl::getPort() const {
+  ::sockaddr_storage addr_s{};
+
+  ::socklen_t addr_len = sizeof(addr_s);
+
+  if (::getsockname(hnd_, reinterpret_cast<::sockaddr*>(&addr_s), &addr_len) != 0) {
+    throw SocketError{"The port number of the socket cannot be retrieved."};
+  }
+
+  if (addr_s.ss_family == AF_INET) {
+    return ntohs(reinterpret_cast<::sockaddr_in*> (&addr_s)->sin_port);
+  } else {
+    return ntohs(reinterpret_cast<::sockaddr_in6*>(&addr_s)->sin6_port);
+  }
+}
+
+bool SocketImpl::setSocketFlag(int lvl, int optname, bool value) noexcept {
+#ifdef _WIN32
+  auto buf = value ? TRUE : FALSE;
+#else
+  auto buf = value ? 1 : 0;
+#endif
+  return setSocketOption(hnd_, lvl, optname, &buf, sizeof(buf)) == 0;
+}
+
+namespace {
+
+struct addrinfo_delete {
+  void operator()(::addrinfo* addr) const noexcept {
+    ::freeaddrinfo(addr);
+  }
+};
+
+using addrinfo_ptr = std::unique_ptr<::addrinfo, addrinfo_delete>;
+
+class SocketListenOp {
+ public:
+  SocketListenOp(std::uint16_t port, const SocketOptions& opts);
+
+  std::unique_ptr<SocketImpl> run();
+
+ private:
+  bool tryListen(int family);
+
+  bool tryListen(const ::addrinfo& addr);
+
+  template <typename... Args>
+  void recordError(fmt::string_view format, Args&&... args) {
+    auto msg = fmt::format(format, std::forward<Args>(args)...);
+
+    logWarning(msg);
+
+    errors_.emplace_back(std::move(msg));
+  }
+
+  std::string port_;
+  const SocketOptions* opts_;
+  std::vector<std::string> errors_{};
+  std::unique_ptr<SocketImpl> socket_{};
+};
+
+SocketListenOp::SocketListenOp(std::uint16_t port, const SocketOptions& opts)
+    : port_{fmt::to_string(port)}, opts_{&opts} {}
+
+std::unique_ptr<SocketImpl> SocketListenOp::run() {
+  if (opts_->prefer_ipv6()) {
+    logInfo("The server socket will attempt to listen on an IPv6 address.");
+    if (tryListen(AF_INET6)) {
+      return std::move(socket_);
+    }
+
+    logInfo("The server socket will attempt to listen on an IPv4 address.");
+    if (tryListen(AF_INET)) {
+      return std::move(socket_);
+    }
+  } else {
+    logInfo("The server socket will attempt to listen on an IPv4 or IPv6 address.");
+    if (tryListen(AF_UNSPEC)) {
+      return std::move(socket_);
+    }
+  }
+
+  constexpr auto* msg = "The server socket has failed to listen on any local network address.";
+
+  logError(msg);
+
+  throw SocketError{fmt::format("{} {}", msg, fmt::join(errors_, " "))};
+}
+
+bool SocketListenOp::tryListen(int family) {
+  ::addrinfo hints{}, *naked_result = nullptr;
+
+  hints.ai_flags = AI_PASSIVE | AI_NUMERICSERV;
+  hints.ai_family = family;
+  hints.ai_socktype = SOCK_STREAM;
+
+  int r = ::getaddrinfo(nullptr, port_.c_str(), &hints, &naked_result);
+  if (r != 0) {
+    const char* gai_err = ::gai_strerror(r);
+
+    recordError("The local {}network addresses cannot be retrieved (gai error: {} - {}).",
```

Comment:
nit: `{}network` -> `{} network` ?

Reply:
It is in fact `{}network` and the possible strings are `IPv4 `, `IPv6 `, and ``. Because we need to have an empty string, a space between `{}` and `network` would cause double space.